src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -747,6 +748,66 @@ private bool CheckFieldValueKind(SyntaxNode node, BoundFieldAccess fieldAccess,
             return CheckIsValidReceiverForVariable(node, fieldAccess.ReceiverOpt, valueKind, diagnostics);
         }
 
+        bool IsAllowedInitOnlyAssignment(Symbol accessedMember, BoundExpression receiver, bool restrictToThis)
+        {
+            // ok: new C() { InitOnlyProperty = ... }
+            if (receiver is BoundObjectOrCollectionValuePlaceholder)
+            {
+                return true;
+            }
+
+            // bad: other.InitOnlyProperty = ...
+            if (!(receiver is BoundThisReference) && !(receiver is BoundBaseReference) && !(receiver is BoundObjectOrCollectionValuePlaceholder))
+            {
+                return false;
+            }
+
+            var containingMember = ContainingMemberOrLambda;
+            if (!(containingMember is MethodSymbol method))
+            {
+                return false;
```

Comment:
>return false [](start = 16, length = 12)

I this going to break an assignment to a redonly static field within another static initializer? Please add a test.

Reply:
When it comes to fields, the purpose of `IsAllowedInitOnlyAssignment` is to relax some assignments to readonly fields. For code that isn't in a method, the assignment shouldn't be relaxed. I can add a test

---
In reply to: [411008763](https://github.com/dotnet/roslyn/pull/43275#discussion_r411008763) [](ancestors = 411008763)

Reply:
> When it comes to fields, the purpose of IsAllowedInitOnlyAssignment is to relax some assignments to readonly fields. For code that isn't in a method, the assignment shouldn't be relaxed. I can add a test

This isn't a relaxation, the scenario is allowed today and we should continue doing so. For example, this is fine:
```
class Test
{
    static readonly int x;
    static int y = (x = 20);
}
```

---
In reply to: [411493817](https://github.com/dotnet/roslyn/pull/43275#discussion_r411493817) [](ancestors = 411493817,411008763)

Reply:
This will still work because the logic for fields is:
```
canModifyReadonly = (containingMethod.MethodKind == desiredMethodKind) ||
                                IsAllowedInitOnlyAssignment(fieldSymbol, fieldAccess.ReceiverOpt, restrictToThis: true); // this can only relax assignment
```

---
In reply to: [411520839](https://github.com/dotnet/roslyn/pull/43275#discussion_r411520839) [](ancestors = 411520839,411493817,411008763)

Reply:
> This will still work because the logic for fields is:

Since the logic is now in a different place, relying on someone else to do part of the work is fragile. The method should be responsible to do the right thing itself, at least it should assert that it is called only under expected conditions (this isn't a local function) and probably the conditions should be documented. Also, I think the code for readonly fields and other members should be split for clarity because rules are quite different. Perhaps some portions could be shared instead.


---
In reply to: [411524374](https://github.com/dotnet/roslyn/pull/43275#discussion_r411524374) [](ancestors = 411524374,411520839,411493817,411008763)